### PR TITLE
Updated Version Format

### PR DESCRIPTION
**Pull request by paulrolfe**
**Summary**
Making the Version type act more like semver.org recommends. While leaving some wiggle room for some common mistakes when it comes to versioning. The typical errors you might see which we would allow here are:
- Leading zeroes are allowed.
  - A common error that should not prevent initialization. 
  - i.e. `1.2.04` should be permitted but translated to it's more correct format: `1.2.4`
- Non-alphanumeric characters are allowed in the `preRelease` and `metaData`
  - Found this requirement to be too cumbersome, when they are just strings anyway.
- Version numbers beyond the patch are dropped (but still create a valid version)
  - For example `1.2.3.4` would still intialize, but it would become `1.2.3`

See http://semver.org for reference.

**Why this is needed**
With the old implementation the following issues arose:
- `1.2.3-beta+abcdef` would have returned a nil version.
- You could not manipulate a version once created.

**What other changes were made**
- New `ip_suffix(from: String)` and `ip_prefix(upTo: String)` methods added
  - This helps in parsing versions from strings.
  - Tests included.
- Slightly updated Comparable rules, as Version has some quirks. Such as `metaData` should not factor into precedence (but `preRelease` should) and the fact that `4.0.0` > `4.0.0-beta2` > `4.0.0-beta`

### CODE REVIEW

**Comment by tzm41** [comment]
Tested on Xcode 9 as well?

**Comment by paulrolfe** [comment]
@tzm41  No, this is an update using swift 3. Swift 4 conversion can happen later -- this particular change is needed in a swift 3 project.

**Comment by IntrepidBuildServer** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/Array+UtilitiesTests.swift#L100">Array+UtilitiesTests.swift#L100</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 3] == 4)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/Array+UtilitiesTests.swift#L101">Array+UtilitiesTests.swift#L101</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 8] == 9)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/Array+UtilitiesTests.swift#L102">Array+UtilitiesTests.swift#L102</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: -1])
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/Array+UtilitiesTests.swift#L103">Array+UtilitiesTests.swift#L103</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/7bc59b9d89a15f0531b194845a457131804ee2e2/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: 9])
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 167 tests, with 0 failures (0 unexpected) in 0.465 (0.529) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by tzm41** [comment]
@paulrolfe We have already updated Swift Wisdom to support Swift 3.2 / Xcode 9: https://github.com/IntrepidPursuits/swift-wisdom/pull/140. And for example BMAP Swift 3.2 will be dependent on Swift Wisdom being Xcode 9 compatible. Just want to make sure that this PR doesn't break Xcode 9 compatibility.

**Comment by paulrolfe** [comment]
@tzm41 Ah, understood. Didn't realize we were in 3.2 already.

**Comment by paulrolfe** [comment]
@tzm41 Confirmed that it works in Xcode9. ✅

**Comment by tzm41** [comment]
@paulrolfe awesome! Thanks for checking!

**Comment by IntrepidBuildServer** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/Array+UtilitiesTests.swift#L100">Array+UtilitiesTests.swift#L100</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 3] == 4)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/Array+UtilitiesTests.swift#L101">Array+UtilitiesTests.swift#L101</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 8] == 9)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/Array+UtilitiesTests.swift#L102">Array+UtilitiesTests.swift#L102</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: -1])
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/Array+UtilitiesTests.swift#L103">Array+UtilitiesTests.swift#L103</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/f930a02648650a7576f56d2b1e6847573f42cf19/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: 9])
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 167 tests, with 0 failures (0 unexpected) in 0.729 (0.863) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Review by alexpersian — APPROVED** [review]

**Comment by IntrepidBuildServer** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/Array+UtilitiesTests.swift#L100">Array+UtilitiesTests.swift#L100</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 3] == 4)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/Array+UtilitiesTests.swift#L101">Array+UtilitiesTests.swift#L101</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 8] == 9)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/Array+UtilitiesTests.swift#L102">Array+UtilitiesTests.swift#L102</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: -1])
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/Array+UtilitiesTests.swift#L103">Array+UtilitiesTests.swift#L103</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c763799f29c0f976becd036171c4166d92c0177a/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: 9])
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 167 tests, with 0 failures (0 unexpected) in 0.610 (0.693) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Review by tzm41 — APPROVED** [review]

File: SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift [hunk 1]
```diff
@@ -79,4 +79,19 @@ extension String {
     public var ip_fullrange: NSRange {
         return NSRange(location: 0, length: ip_length)
     }
+
+    /// Returns a new string containing the characters of the String up to, but not including, the first occurence of the given string.
```
Comment (bobgilmore) [inline SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift]: "Occurrence" has two "r"s.

File: SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift [hunk 2]
```diff
@@ -79,4 +79,19 @@ extension String {
...
+    /// If the string is not found, nil is returned.
+    public func ip_prefix(upTo string: String) -> String? {
+        guard let range = self.range(of: string) else { return nil }
+        return self.substring(to: range.lowerBound)
+    }
+
+    /// Returns a new string containing the characters of the String from the end of the first occurence of the given String.
```
Comment (bobgilmore) [inline SwiftWisdom/Core/StandardLibrary/String/String+Indexing.swift]: "Occurrence" has two "r"s.

**Review by bobgilmore — CHANGES_REQUESTED** [review]
Very minor changes to the doc comments.

**Review by bobgilmore — APPROVED** [review]

**Comment by ghost** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/Array+UtilitiesTests.swift#L100">Array+UtilitiesTests.swift#L100</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L100</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 3] == 4)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/Array+UtilitiesTests.swift#L101">Array+UtilitiesTests.swift#L101</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L101</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssert(testArray[ip_safe: 8] == 9)
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/Array+UtilitiesTests.swift#L102">Array+UtilitiesTests.swift#L102</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L102</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: -1])
</code></td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/Array+UtilitiesTests.swift#L103">Array+UtilitiesTests.swift#L103</a> - <strong><a href="https://github.com/IntrepidPursuits/swift-wisdom/blob/c28e4d552d8199c1e76648db43641f562db8203e/SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103">SwiftWisdomTests/StandardLibrary/Array/Array+UtilitiesTests.swift#L103</a></strong>: ‘subscript(ip_safe:)’ is deprecated: Use <code>ip_safely</code> instead.  <br /><code>
        XCTAssertNil(testArray[ip_safe: 9])
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 167 tests, with 0 failures (0 unexpected) in 0.605 (0.688) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Pull request merged**